src/network/socket/Socket.hh

```diff
@@ -0,0 +1,33 @@
+<?hh // strict
+
+namespace HH\Lib\Experimental\Network;
+
+use type Throwable;
+
+interface Socket {
+  /**
+   * Close the underlying socket.
+   * 
+   * @param Throwable $e Reason for close, will be set as previous error.
+   */
+  public function close(?Throwable $e = null): void;
+
+  /**
+   * Get the local address of the socket.
+   */
+  public function getAddress(): string;
+
+  /**
+   * Get the local network port, or NULL when no port is being used.
+   */
+  public function getPort(): ?int;
+
+  /**
+   * Change the value of a socket option, options are declared as class constants.
+   * 
+   * @param int $option Option to be changed.
+   * @param mixed $value New value to be set.
+   * @return bool Will return false when the option is not supported.
+   */
+  public function setOption(int $option, mixed $value): bool;
```

Comment:
where are these options defined?
- should this be an enum?
- is this derived from some other implementaiton? if so, please add original license headers/information

Reply:
since there's no shared options between `TcpSocket` and `UdpSocket` im removing the `setOption` method from `Socket` and refactoring the `TcpSocket` and `UdpSocket` to use `shape()`

Reply:
see https://github.com/hhvm/hsl-experimental/pull/36/commits/b72067e2832d84e7b68fbd4f9fefc00cfe459ad6